File: Fortran/One_sided/Accumulate/accumulate_start_post.f90
PROGRAM accumulate

  USE mpi
  IMPLICIT NONE

  INTEGER, PARAMETER :: tag = 100

  INTEGER :: rank, src, dest, nproc, dummy
  INTEGER :: comm_group, zero_group
  INTEGER :: ierr, window, intsize
  INTEGER(KIND=MPI_ADDRESS_KIND) :: size_of_window, offset

  CALL MPI_Init(ierr)

  CALL MPI_Comm_size(MPI_COMM_WORLD, nproc, ierr)
  CALL MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr)

  !In C would just use sizeof()
  !In Fortran2008, could use C_SIZEOF
  !NOTE! This routine returns the size of AN ELEMENT
  !Not the size of an array if you give it an array
  CALL MPI_Sizeof(src, intsize, ierr)
  !Just using a single int here
  size_of_window = intsize

  !Create the window. This is a piece of memory that's available for remote
  !access. In this case, a single 4 byte integer
  CALL MPI_Win_create(dest, size_of_window, intsize, MPI_INFO_NULL, &
      MPI_COMM_WORLD, window, ierr)

  CALL MPI_Comm_group(MPI_COMM_WORLD, comm_group, ierr)
  CALL MPI_Group_incl(comm_group, 1, (/0/), zero_group, ierr)

  IF (rank == 0) CALL MPI_Win_post(comm_group, 0, window, ierr)
  CALL MPI_Win_start(zero_group, 0, window, ierr)

  src = rank
  dest = 0

  !Put the result into the first (zeroth) slot
  offset = 0

  !This call accumulates whatever data is in src on all processors and puts
  !them in the window on processor 0. Here using MPI_Win_post and MPI_Win_start
  !to put things in the most restrictive epoch possible
  CALL MPI_Accumulate(src, 1, MPI_INTEGER, 0, offset, 1, MPI_INTEGER, &
      MPI_SUM, window, ierr)

  CALL MPI_Win_complete(window, ierr)
  IF (rank == 0) CALL MPI_Win_wait(Window, ierr)


  IF (rank == 0) PRINT *,"Accumulated value is ", dest

  CALL MPI_Finalize(ierr)

END PROGRAM accumulate

File: C/One_sided/Accumulate/accumulate_start_post.c
#include <stdio.h>
#include <mpi.h>

#define TAG 100

int main(int argc, char ** argv)
{

  int rank, nproc, src, dest, ranks;
  MPI_Aint size_of_window, offset;
  MPI_Win window;
  MPI_Group comm_group, zero_group;

  MPI_Init(&argc, &argv);
  MPI_Comm_size(MPI_COMM_WORLD, &nproc);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);

  offset = 0;
  size_of_window = sizeof(int);

  //Create the window. This is a piece of memory that's available for remote
  //access. In this case, a single 4 byte integer
  MPI_Win_create(&dest, size_of_window, sizeof(int), MPI_INFO_NULL,
      MPI_COMM_WORLD, &window);

  ranks = 0;
  MPI_Comm_group(MPI_COMM_WORLD, &comm_group);
  MPI_Group_incl(comm_group, 1, &ranks, &zero_group);

  //Zero is where data is being put, so only 0 needs to call MPI_Win_post
  if (rank == 0) MPI_Win_post(comm_group, 0, window);
  //Because zero is writing to itself, it also has to call MPI_Win_start
  MPI_Win_start(zero_group, 0, window);

  //Put the result into the first (zeroth) slot
  offset = 0;
  //Accumulate ranks and zero destination
  src = rank;
  dest = 0;

  MPI_Accumulate(&src, 1, MPI_INT, 0, offset, 1, MPI_INT,
      MPI_SUM, window);

  MPI_Win_complete(window);
  if (rank == 0) MPI_Win_wait(window);

  if (rank == 0) printf("Accumulated value is %3d\n", dest);

  MPI_Finalize();

}
